### PR TITLE
[DRAFT] RFC 1008 PoC: TLS and ALPN

### PR DESCRIPTION
**Pull request by fantix**
Draft for PoC purposes, please don't waste time reviewing yet.

Refs edgedb/rfcs#37

### CODE REVIEW

File: edb/server/args.py [hunk 1]
```diff
@@ -317,6 +321,23 @@ def _writer(status: str) -> None:
...
+             'signed certificate will be generated and used instead (for '
+             'development only!).'),
+    click.option(
+        '--tls-keyfile', type=PathPath(),
+        help='Specify a path to a file containing the private key. If not '
+             'present, the private key will be taken from --tls-certfile as '
+             'well. If the private key is protected by a password, specify '
+             'with an environment variable EDGEDB_TLS_PRIVATE_KEY_PASSWORD.'),
+    click.option(  # TODO: make the default False
+        '--allow-no-tls', type=bool, default=True, is_flag=True, hidden=True,
```
Comment (1st1) [inline edb/server/args.py:338]: This one is only for `edb server` and not for `edgedb-server`? I think these options are shared between the two

File: edb/server/protocol/protocol.pyx [hunk 2]
```diff
@@ -75,25 +79,50 @@ cdef class HttpProtocol:
...
+                outgoing = ssl.MemoryBIO()
+                incoming = ssl.MemoryBIO()
+                incoming.write(data)
+                sslobj = self.server._sslctx.wrap_bio(
+                    incoming, outgoing, server_side=True
+                )
+                sslobj.do_handshake()
+            except ssl.SSLWantReadError:
+                pass
+            except Exception:
```
Comment (1st1) [inline edb/server/protocol/protocol.pyx:101]: Probably we should use less genetic exception type here

File: edb/server/protocol/protocol.pyx [hunk 3]
```diff
@@ -75,25 +79,50 @@ cdef class HttpProtocol:
...
                 # This is, most likely, our binary protocol,
                 # as its first message kind is `V`.
                 #
                 # Switch protocols now (for compatibility).
-                binproto = binary.EdgeConnection(
-                    self.server, self.external_auth)
-                self.transport.set_protocol(binproto)
-                binproto.connection_made(self.transport)
-                binproto.data_received(data)
+                self._switch_to_binary_protocol(data)
```
Comment (1st1) [inline edb/server/protocol/protocol.pyx:121]: Only when `if self.server._tls_compat`?

**Review by 1st1 — COMMENTED** [review]
Nice!

File: edb/server/protocol/protocol.pyx (hunk 3 above)
Comment (fantix) [reply edb/server/protocol/protocol.pyx:121]: My bad ... this is actually needed by `_return_binary_error()` - I'll make it more readable.

File: edb/server/args.py (hunk 1 above)
Comment (fantix) [reply edb/server/args.py:338]: Yes you're right, `edgedb-server` has these options too. At least `edgedb server` the CLI won't have it.

**Pull request closed without merging**